3scale/porta
# THREESCALE-10721: Fix Liquid error when rendering a custom form with reCAPTCHA v3

**What this PR does / why we need it**:

I'm not specially proud of this, but I create this PR to get feedback and launch tests.

The problem happens when trying to render a Liquid protected form (those including `ThreeScale::BotProtection::Form`), e.g. the [login form](https://github.com/3scale/porta/blob/2436ac7d83cc98338719dc40183983649d306e4b/lib/developer_portal/lib/liquid/forms/bot_protected.rb).

```liquid
{% form 'login', user, class: 'form-horizontal' %}
  <fieldset>
  ...    
  </fieldset>
{% endform %}
```

When rendering this from a path other than the builtin login one, it invokes `DeveloperPortal::CMS::NewContentController` instead of `DeveloperPortal::LoginController`. These two controllers follow a different execution path to end up rendering the liquid template. And in the case of `NewContentController`, it ends up calling `template` when `context.registers[:view]` hasn't been initialized yet, [here](https://github.com/3scale/porta/blob/master/lib/developer_portal/lib/liquid/forms/bot_protected.rb#L31).

I've been trying to understand the rendering process to figure out a way to ensure `context.registers[:view]` is initialized at [this point](https://github.com/3scale/porta/blob/c38c9e1688b1824188487b1b1dfe9014b760ffcb/lib/developer_portal/app/controllers/developer_portal/cms/new_content_controller.rb#L57), which in my opinion would fix the problem. But the code is so obfuscated I lost track and ended up just writing a workaround to ensure `template` is not called at all.

What I don't like of this solution is `ThreeScale::BotProtection::Form` should be self-contained and able to be included from wherever, but instead I had to make changes in `Liquid::Forms::BotProtected` and `ThreeScale::SemanticFormBuilder` in order to met some conditions so `ThreeScale::BotProtection::Form` can be properly included. Those conditions are having `:controller` and `:site_account` defined. If we want to include `ThreeScale::BotProtection::Form` to some other module in the future, we'll have to remember that.

Any better ideas?

**Which issue(s) this PR fixes** 

[THREESCALE-10721](https://issues.redhat.com/browse/THREESCALE-10721)

**Verification steps**

You'll need to verify the recaptcha presence in the HTML source at least from this three locations:

1. The login screen in the Developer portal: http://provider.3scale.localhost:3000/login
2. A custom Developer portal template including the login form.
3. The provider signup screen for master http://master-account.3scale.localhost:3000/p/signup

**Additional notes**

Not related to the PR but worth mentioning. During the investigation I found this:

https://github.com/3scale/porta/blob/a40360acde0c9d717364c42877273f69806f8fc6/app/lib/liquid/template_support.rb#L51-L66

When a controller calls `liquify`, it adds a `before_action` [callback](https://github.com/3scale/porta/blob/a40360acde0c9d717364c42877273f69806f8fc6/app/lib/liquid/template_support.rb#L20-L22) that overwrites this Rails method once for each request we receive in any controller calling `liquify`. It looks like we should move this to an initializer, as the comment suggests.
